TaylorSource/CoreData/NSFRCUpdateHandler.swift

```diff
@@ -0,0 +1,115 @@
+import UIKit
+import CoreData
+
+private class WeakObserver {
+    private(set) weak var value: NSFRCIndexedUpdateConsumer?
+
+    init(_ observer: NSFRCIndexedUpdateConsumer) {
+        value = observer
+    }
+}
+
+public class NSFRCUpdateHandler: NSObject, NSFetchedResultsControllerDelegate {
+    
+    deinit {
+        self.observers.removeAll()
+    }
+    
+    public func addUpdateObserver(observer: NSFRCIndexedUpdateConsumer) {
+        self.observers.append(WeakObserver(observer))
+    }
+    
+    func fullUpdate() {
+        self.sendUpdate(.FullUpdate)
+    }
+    
+    // MARK: Private
+    
+    private var observers = [WeakObserver]()
+    
+    private func sendUpdate(update: NSFRCIndexedUpdate) {
+        self.observers = self.observers.filter { $0.value != nil } // Remove orphaned observers
+        self.observers.forEach { $0.value?.handleIndexedUpdate(update) }
+    }
+    
+    private func createUpdateFromCollectedValues() -> NSFRCIndexedUpdate {
+        let insertedSections = self.insertedSections.copy() as! NSIndexSet
+        let deletedSections = self.deletedSections.copy() as! NSIndexSet
+        let update: NSFRCIndexedUpdate = .DeltaUpdate(
+            insertedSections: insertedSections,
+            deletedSections: deletedSections,
+            insertedRows: self.insertedRows,
+            updatedRows: self.updatedRows,
+            deletedRows: self.deletedRows
+        )
+        return update
+    }
+    
+    private func clearCollectedValues() {
+        self.insertedSections = nil
+        self.deletedSections = nil
+        self.insertedRows = nil
+        self.updatedRows = nil
+        self.deletedRows = nil
+    }
+    
+    // MARK: NSFetchedResultsControllerDelegate
+    
+    private var insertedSections: NSMutableIndexSet!
+    private var deletedSections: NSMutableIndexSet!
+    private var insertedRows: [NSIndexPath]!
+    private var updatedRows: [NSIndexPath]!
+    private var deletedRows: [NSIndexPath]!
+    
+    public func controllerWillChangeContent(controller: NSFetchedResultsController) {
+        self.insertedSections = NSMutableIndexSet()
+        self.deletedSections = NSMutableIndexSet()
+        self.insertedRows = []
+        self.updatedRows = []
+        self.deletedRows = []
+    }
```

Comment:
This strikes me as a little bit of a code smell.

We've got this `enum` which encapsulates all the changes we need the view to update with. And there is this class which acts as the NSFRC delegate. It's job is collect those changes.

However, I think we need another type which will be responsible for storing the changes and creating the enum. This would mean we can remove these explicitly unwrapped types from here, and instead write an immutable struct which works like this:

```swift
struct PendingChanges {
    let insertedSections: Set<Int>
    let removedSections: Set<Int>

   // etc etc

    func addSectionIndex(index: Int) -> PendingChanges {
       var newInsertedSections = insertedSections
       newInsertedSections.insert(index)
       return PendingChanges(insertedSections: newInsertedSections, removedSections: removedSections)
   }

    // etc etc

    func createUpdate() -> NSFRCIndexedUpdate {
        // etc etc
    }
}
```

This means you can have tests which for how that enum is created, and consumed without having to drive it all through the NSFetchedResultsControllerDelegate API.

Also, I'd look into using `Set<Int>` instead of NSIndexSet to store these things - as it's just so much nicer to work a real (im)mutability programming model.

Given all of this, this means that your property looks like this:

```swift
var pendingChanges = PendingChanges() // i.e. empty change set - better than explicitly unwrapped or Optional.
```

```swift
public func controllerWillChangeContent(controller: NSFetchedResultsController) {
    // Reset the pending changes...
    pendingChanges = PendingChanges()
}
```

Thoughts? 🤔